Add Mock Datastores

tarzzz opened the pull request:
This PR adds the mock datastore which can be used for testing the falcon API.

To enable the mock datastore, we need to start falcon as:

```
PLOTLY_CONNECTOR_AUTH_ENABLED=false PLOTLY_CONNECTOR_TEST_MODE=true yarn run start-headless
```

cc: @n-riesco @scjody @chriddyp @bpostlethwaite

tarzzz commented:
I have added the generic mocks for SQL-Like Databases and Elasticsearch. 
Currently working on other databases (Apache Drill, S3 etc.)..

tarzzz commented:
@chriddyp or @n-riesco ready for your review.

  - These mocks handle all the connections we currently support.
  - I have added the minimum required hardcoded configurations for our purposes.
  - The connection strings will still be picked from connections.yaml as before (i.e. pointing to test DBs), but the query (and sql-tables) results will be mocked. 
  - For handling syntax-error cases, I have added a specific case where `ERROR` as query string returns a Promise Reject/syntax error.

n-riesco commented on backend/persistent/datastores/DataStoreMock.js:
```diff
@@ -0,0 +1,147 @@
+}
+
+export function query(queryString) {
+
```
:cow:

n-riesco commented on backend/persistent/datastores/DataStoreMock.js:
```diff
@@ -0,0 +1,147 @@
+export function query(queryString) {
+
+    return new Promise((resolve, reject) => {
+
```
:cow:

n-riesco commented on backend/persistent/datastores/DataStoreMock.js:
```diff
@@ -0,0 +1,147 @@
+}
+
+export function files() {
+    return new Promise((resolve) => {
```
Rewrite as `return Promise.resolve([...]);`

n-riesco commented on backend/persistent/datastores/DataStoreMock.js:
```diff
@@ -0,0 +1,147 @@
+}
+
+export function elasticsearchMappings() {
+    return new Promise((resolve) => {
```
Rewrite as `return Promise.resolve([...]);`

n-riesco commented on backend/persistent/datastores/DataStoreMock.js:
```diff
@@ -0,0 +1,147 @@
+}
+
+export function storage() {
+
```
:cow:

n-riesco commented on backend/persistent/datastores/DataStoreMock.js:
```diff
@@ -0,0 +1,147 @@
+
+export function storage() {
+
+    return new Promise((resolve) => {
```
Rewrite as `return Promise.resolve([...]);`

n-riesco commented on backend/persistent/datastores/DataStoreMock.js:
```diff
@@ -0,0 +1,147 @@
+
+// This is for Apache Drill:
+export function listS3Files() {
+  return new Promise((resolve) => {
```
Rewrite as `return Promise.resolve([...]);`

n-riesco commented on backend/persistent/datastores/Datastores.js:
```diff
@@ -5,7 +5,9 @@ import * as ApacheDrill from './ApacheDrill';
 import * as IbmDb2 from './ibmdb2';
 import * as ApacheLivy from './livy';
 import * as ApacheImpala from './impala';
+import * as DataStoreMock from './DataStoreMock';
```
`%s/DataStore/datastore/g`

n-riesco replied:
use lower case filenames (otherwise, when developing in mac, it's very easy to commit filename typos)

n-riesco commented on backend/persistent/datastores/Datastores.js:
```diff
@@ -27,6 +29,12 @@ import * as ApacheImpala from './impala';
     const {dialect} = connection;
+
+    // handle test mode:
+    if (getSetting('TEST_MODE')) {
```
can this be done without introducing a new ENV var?

tarzzz replied:
I am not sure of the other way. We would typically need to enable this from outside the container, ie. when we run the container. 

```
PLOTLY_CONNECTOR_TEST_MODE=true docker run -p 9494:9494 falcon
```

Please let me know if you know of any other way to accomplish the same !!

n-riesco commented on backend/persistent/datastores/Datastores.js:
```diff
@@ -27,6 +29,12 @@ import * as ApacheImpala from './impala';
+
+    // handle test mode:
+    if (getSetting('TEST_MODE')) {
+        return DataStoreMock;
```
`%s/DataStore/datastore/g`

n-riesco commented on backend/settings.js:
```diff
@@ -50,6 +50,11 @@ const DEFAULT_SETTINGS = {
+     * If TEST_MODE is true, Falcon oes not make requests to the
+     * databases, but instead return hardcoded responses.
+     */
+    TEST_MODE: false,
```
can this be done without introducing a new ENV var?

n-riesco commented:
Here's my review:
- [ ] Replace `DataStore` with `Datastore` to be consistent with the existing code
- [ ] Use lowercase filenames (otherwise, when developing in mac, it's very easy to commit filename typos)
- [ ] Add tests
- [ ] Please, add a git hook to your setup, so that nothing that fails eslint gets committed

Questions:
- I don't see how this PR can be tested using the current specs
- It also looks like this PR will require changes in Workspace (since the mocked connectors return values  different to the real connectors)
- Can this PR be implemented without introducing a new ENV var? We have so many ENV vars that the code in Falcon is becoming unwieldy.

I don't like the idea of coupling the connectors with settings. Until now connectors don't use settings.

n-riesco commented:
> I am not sure of the other way. We would typically need to enable this from outside the container, ie. when we run the container.
>
> `PLOTLY_CONNECTOR_TEST_MODE=true docker run -p 9494:9494 falcon`
>
> Please let me know if you know of any other way to accomplish the same !!

How about using `PLOTLY_CONNECTOR_QUERIES_PATH`?

Instead of pointing to the sample connections (`sample-storage/connections.yaml`), make it point to the new mock connections (e.g. `sample-storage/connections.yaml`).

Remember that we discussed the use of the property `connection.mock` to flag a mock connection. If you do that, you'd also address one of the concerns I raised above (*"I don't like the idea of coupling the connectors with settings. Until now connectors don't use settings"*).

tarzzz commented:
> I don't see how this PR can be tested using the current specs

Added new tests for this feature

> It also looks like this PR will require changes in Workspace (since the mocked connectors return values different to the real connectors)

Yes I will open a PR to update the Workspace tests. I have started working on the same.

> Can this PR be implemented without introducing a new ENV var? We have so many ENV vars that the code in Falcon is becoming unwieldy.

Added the updates for the same..

Please let me know if it looks good.

n-riesco commented:
> > I don't see how this PR can be tested using the current specs
>
> Added new tests for this feature

The tests in https://github.com/plotly/falcon-sql-client/pull/329/commits/5108f0ec5aecc6f6afa5775284f3c287c178dbdf aren't really very effective because they call the mocked functions directly.

Ideally, what we want is to test the results returned by Falcon's backend.

n-riesco commented:
And about https://github.com/plotly/falcon-sql-client/pull/329/commits/f98bd43e03cd50dc93acccf493acadb25f67a4f0 :

We want to keep `sample-storage` as is because it's used by the heroku app.

I'd create a new folder, e.g `mock-storage`, and I'd only add as many connectors as you want to test in Workspace.

tarzzz commented:
@n-riesco I will update the tests and add mock connections separately.

Actually on a second thought, I think we should be keeping filenames and Datastore names uniform for all the DataStores, so I feel it is better to keep the name `DatastoreMock.js`. like `ApacheDrill.js`, `Elasticsearch.js` etc..

n-riesco commented:
@tarzzz 

> I will update the tests and add mock connections separately.

No, please, do it in this PR. This PR won't be merged without tests.

> Actually on a second thought, I think we should be keeping filenames and Datastore names uniform for all the DataStores, so I feel it is better to keep the name DatastoreMock.js. like ApacheDrill.js, Elasticsearch.js etc..

No, please, files with lower and upper case are the exception in this project. And I intend to rename them to lower case in the near future.

Personally, I would've named the file `mock.js`.

tarzzz commented:
@n-riesco Updated as per the comments..

n-riesco commented on sample-storage/mock-connections.yaml:
```diff
@@ -0,0 +1,115 @@
+# Credentials to the test databases.
```
This file should be in a different folder and it must be named `connections.yaml`, otherwise `getQueries` won't load the file.

See my comment for `test/backend/datastores.mock.spec.js`.

n-riesco commented on test/backend/datastores.mock.spec.js:
```diff
@@ -0,0 +1,183 @@
+        servers = createTestServers();
+        saveSetting('AUTH_ENABLED', false);
+        sqlConnections.mock = true;
+        connectionId = saveConnection(sqlConnections);
```
You don't want to do this (the above two lines will overwrite `sample-storage/connections.yaml`).

I can't test this until Thursday, but I reckon it's enough to create a mock connections file in `mock-storage/connections.yaml` and set `STORAGE_PATH` accordingly:
```
saveSetting('STORAGE_PATH', 'mock-storage');
```

tarzzz replied:
I checked, and it does not. We are good here.

n-riesco replied:
How do you explain the failure in CircleCI?

n-riesco replied:
I don't have the rights to trigger a rebuild. Please, could you do it for me?

n-riesco replied:
To reproduce the error in CircleCI, you can run locally this:
```
NODE_ENV=test BABEL_DISABLE_CACHE=1 electron-mocha --bail --full-trace --timeout 90000 --compilers js:babel-register test/backend/datastores.mock.spec.js test/backend/routes.spec.js
```

n-riesco replied:
Note that `createTestServers` uses `sample-storage/connections.yaml`.

tarzzz replied:
> and set STORAGE_PATH accordingly:

Unfortunately: https://github.com/plotly/falcon-sql-client/blob/master/backend/settings.js#L171

```
        /*
         * STORAGE_PATH is the only variable that can't be saved in the
         * settings file because of recursion.
         */
```

I am going to remove `mock-connections.yaml`, and add a minor update to the original approach to get this working.

n-riesco replied:
There is no need to save `STORAGE_PATH`. Please, have a look at how the heroku app loads the sample connections.

tarzzz replied:
What is wrong with the current approach? It removes the use of an extra file/directory and does not leave any global effects.

n-riesco commented on test/backend/datastores.mock.spec.js:
```diff
@@ -0,0 +1,183 @@
+    afterEach(() => {
+        clearSettings('SETTINGS_PATH');
+        return closeTestServers(servers);
+
```
:goat:

tarzzz replied:
what does 🐐 mean here?

scjody replied:
I'm not sure either.

For reference, here's the list: https://github.com/plotly/dev-docs/blob/master/basics/code-review-checklist.md#code-review-animals

n-riesco replied:
:space_invader:

tarzzz replied:
This is not _too_ much space IMHO. I am not sure if any Javascript guidelines mandates to not leave any blank lines in code. Can you please point to the relevant document..

n-riesco replied:
:disappointed: Let's not waste our time discussing over white space. There are more interesting things to fix in this PR.

n-riesco commented on test/backend/datastores.mock.spec.js:
```diff
@@ -0,0 +1,183 @@
+        .then(json => {
+            assert.deepEqual(json, ['TABLE_A', 'TABLE_B', 'TABLE_C', 'TABLE_D']);
+        });
+
```
:goat:

tarzzz replied:
?

n-riesco replied:
:space_invader:

tarzzz commented:
@n-riesco Added the updates.. Please let me know if it looks good !!

tarzzz commented:
Ok, the original problem resurfaces. We cannot switch the connector to return mock-ed responses from outside.

I can see two ways to do this:

1. Add a mock connections file under `mock-storage/connections.yaml` and reference that from outside using environment variable to the container. This file will not be used for falcon tests-suite, but can be used for testing API from outside.

2. Replace the condition to use the environment variable instead of `mock` property. Which involves adding another environment variable, which @n-riesco isn't onboard with I believe. 

@n-riesco Okay with `1)` ? cc: @scjody 

Apart from that, this PR is good imho..

n-riesco commented:
> Ok, the original problem resurfaces. We cannot switch the connector to return mock-ed responses from outside.

What have you tried?

tarzzz commented:
> What have you tried?

I am sorry I do not understand. I just mentioned two approaches we can use which I can think of. Please let me know if there is any other approach you want to suggest..

n-riesco commented:
> I am sorry I do not understand. I just mentioned two approaches we can use which I can think of. Please let me know if there is any other approach you want to suggest..

Before going for any of the 2 alternative approaches, I need to know why the initial approach doesn't work.

(I'm on the road; I won't be able to look into this until my evening)

tarzzz commented:
> I need to know why the initial approach doesn't work.

Because we removed two initial approaches entirely (due to design considerations):

1. The use of environment variables which could be switched on/off from outside.
2. Adding `mock=true` in `sample-storage/connections.yaml` which was also removed in reviews.

Therefore currently there is no method to use mocks from outside the container, hence we need to use one of https://github.com/plotly/falcon-sql-client/pull/329#issuecomment-354284111 (or if you have any better ideas, which are welcome)

n-riesco commented:
> Because we removed two initial approaches entirely (due to design considerations):
>
>    1. The use of environment variables which could be switched on/off from outside.
>    2. Adding mock=true in sample-storage/connections.yaml which was also removed in reviews.

I think there is a misunderstanding here. None of those options are what we discussed on slack.

The main two ideas for this PR are:

1. When `connection.mock` is set to `true`, `getDatastoreClient` returns a mock connector.

2. Create `mock-storage/connections.yaml` with a small number of mock connections and, to use it, set the ENV variable `PLOTLY_CONNECTOR_STORAGE_PATH` to `mock-storage`.

tarzzz commented:
Added the mock storage.. 

💃 ??

n-riesco commented:
> :dancer: ??

There is still one more thing. The spec `test/backend/datastores.mock.spec.js` should use the connections in `mock-storage/connections.yaml`.

If you want `createTestServers` to use `mock-storage/connections.yaml`, you'll have to backup `process.env.PLOTLY_CONNECTOR_STORAGE_PATH` to `mock-storage` before calling `createTestServers`, and restore `process.env.PLOTLY_CONNECTOR_STORAGE_PATH` after invoking `closeTestServers`.

tarzzz commented:
> There is still one more thing. The spec test/backend/datastores.mock.spec.js should use the connections in mock-storage/connections.yaml.

I don't think that is necessary. Current approach looks fine to me. Unless there is a specific reason we should be doing that, I would like to keep it as such.

n-riesco commented:
> I don't think that is necessary. Current approach looks fine to me. Unless there is a specific reason we should be doing that, I would like to keep it as such.

As it stands `mock-storage/connections.yaml` isn't used.  How do you intend Workspace to use mock connections? Please, let's discussion on slack. I'm closing the PR until we reach an agreement.

tarzzz commented:
> How do you intend Workspace to use mock connections?

We can set `PLOTLY_CONNECTOR_STORAGE_PATH` to "mock-storage" **while running the container from streambed side**. This is the reason why "mock-storage" is included.. 

> As it stands mock-storage/connections.yaml isn't used

Not used in falcon, because we don't need to.This will be used when we are running Falcon as the part of streambed tests, as explained above.

> Please, let's continue the discussion on slack.

I would prefer we do it here..!!

bpostlethwaite commented:
Okay if the PR is good from a streambed point of view (I don't understand all the stuff here) the main remaining ask from @n-riesco is https://github.com/plotly/falcon-sql-client/pull/329#issuecomment-354668231 

Looks like the tests currently use the sample storage instead of the mock storage. Doesn't that mean we are not actually testing:
> We can set PLOTLY_CONNECTOR_STORAGE_PATH to "mock-storage" while running the container from streambed side. This is the reason why "mock-storage" is included..
@tarzzz  ?

Seems like it wouldn't be too much to test this pathway right? @n-riesco anything else before this is mergable? 

@tarzzz anything else?

tarzzz commented:
Updates made:

1. Uses "mock-storage" for mock tests.
2. Added updates to use "mock-storage" for Heroku App (as discussed on slack channel).

@n-riesco Please let me know if it looks good..

n-riesco commented on test/backend/datastores.mock.spec.js:
```diff
@@ -15,19 +15,21 @@ import {
 
 let servers;
-let connectionId;
+let connectionId = 'postgres-189ebfb4-e1b4-446c-9b83-b326875fa2d8';
```
`const`?

n-riesco replied:
is there an easy way to check that all the connections in `mock-storage` have been read?

tarzzz replied:
Updated.. !!

n-riesco commented on test/backend/datastores.mock.spec.js:
```diff
@@ -15,19 +15,21 @@ import {
-        connectionId = saveConnection(sqlConnections);
+        clearSettings('SETTINGS_PATH');
+        process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
+        saveSetting('AUTH_ENABLED', true);
```
Is `saveSetting('AUTH_ENABLED', true);` needed?

tarzzz replied:
I think so. Because we should restore any changes made to settings upon test completion..

n-riesco commented on test/backend/datastores.mock.spec.js:
```diff
@@ -15,19 +15,21 @@ import {
     afterEach(() => {
-        sqlConnections.mock = false;
-        connectionId = saveConnection(sqlConnections);
+        clearSettings('SETTINGS_PATH');
```
Is `clearSettings('SETTINGS_PATH');` needed? (`createTestServers();` does `clearSettings('CONNECTIONS_PATH', 'QUERIES_PATH', 'SETTINGS_PATH');`)

tarzzz replied:
> Is clearSettings('SETTINGS_PATH'); needed? 

Yes, otherwise "mock-storage/settings.yaml" will be created on every test run.

n-riesco replied:
I don't follow the reasoning: `clearSettings('SETTINGS_PATH');` removes `mock-storage/settings.yaml`.

I've just tested `test/backend/datastores.mock.spec.js` without `clearSettings('SETTINGS_PATH');` and `saveSetting('AUTH_ENABLED', true);` and the specs pass.

tarzzz replied:
Ok, Removed the lines..!!

n-riesco commented on test/backend/datastores.mock.spec.js:
```diff
@@ -15,19 +15,21 @@ import {
+        clearSettings('SETTINGS_PATH');
+        process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
+        saveSetting('AUTH_ENABLED', true);
         return closeTestServers(servers);
```
I've just realised that this should be:
```js
return closeTestServers(servers).then(() => {
    process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
});
```
Otherwise `closeTestServers` will use the wrong `STORAGE_PATH`.

tarzzz replied:
Updated .. !!

n-riesco commented:
@tarzzz I've noticed that after running `test/backend/datastores.mock.spec.js` a folder named `undefined` and file `mock-storage/settings.yaml` are left behind.

This is caused by [beforeEach](https://github.com/plotly/falcon-sql-client/blob/0621fa84104f34d22b7128347bbb36547a2d4b69/test/backend/datastores.mock.spec.js#L20-L25) using `createTestServers()`.

I think the easiest solution is to create `Servers` directly. The following works for me:

```diff
user@host:~/github/plotly-database-connector$ git diff HEAD~1
diff --git a/test/backend/datastores.mock.spec.js b/test/backend/datastores.mock.spec.js
index 9bfca2e..35ea419 100644
--- a/test/backend/datastores.mock.spec.js
+++ b/test/backend/datastores.mock.spec.js
@@ -1,13 +1,14 @@
 import {assert} from 'chai';
 
+import Servers from '../../backend/routes.js';
 import {saveSetting} from '../../backend/settings.js';
 
 import {
     assertResponseStatus,
     clearSettings,
     closeTestServers,
-    createTestServers,
     getResponseJson,
+    GET,
     POST
 } from './utils.js';
 
@@ -19,17 +20,28 @@ const connectionId = 'postgres-189ebfb4-e1b4-446c-9b83-b326875fa2d8';
 
 describe('Datastore Mock:', function () {
     beforeEach(() => {
-        servers = createTestServers();
         storagePath = process.env.PLOTLY_CONNECTOR_STORAGE_PATH;
         process.env.PLOTLY_CONNECTOR_STORAGE_PATH = 'mock-storage';
+
         saveSetting('AUTH_ENABLED', false);
+
+        servers = new Servers({createCerts: false, startHttps: false, isElectron: false});
+        servers.httpServer.start();
     });
 
     afterEach(() => {
-        clearSettings('SETTINGS_PATH');
-        process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
-        saveSetting('AUTH_ENABLED', true);
-        return closeTestServers(servers);
+        return closeTestServers(servers).then(() => {
+            clearSettings('SETTINGS_PATH'); // otherwise `mock-storage/settings.yaml` is left behind
+            process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
+        });
+    });
+
+    it('all mock connections are loaded', function() {
+        return GET('connections')
+        .then(assertResponseStatus(200)).then(getResponseJson)
+        .then(json => {
+            assert.equal(Object.keys(json).length, 12, 'Missing mock connectors');
+        });
     });
```

tarzzz commented:
@n-riesco: Addressed comments..

n-riesco commented:
@tarzzz I've just found out why the folder `undefined` is created. we should replace:
```js
process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
```
by:
```js
if (storagePath) process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
```

n-riesco commented on test/backend/datastores.mock.spec.js:
```diff
@@ -32,7 +32,9 @@ describe('Datastore Mock:', function () {
             clearSettings('SETTINGS_PATH');
-            process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
+            if (storagePath) {
+                process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
```
we should delete `process.env.PLOTLY_CONNECTOR_STORAGE_PATH` if it wasn't defined initially.

tarzzz replied:
By delete, you mean set it to `null`?

tarzzz replied:
> if it wasn't defined initially.

What does this means exactly? We do define it initially..

tarzzz replied:
Would be great if you could post a code snippet of what you are trying to say.. :)

n-riesco replied:
```js
if (storagePath) process.env.PLOTLY_CONNECTOR_STORAGE_PATH = storagePath;
else delete process.env.PLOTLY_CONNECTOR_STORAGE_PATH;
```

tarzzz commented:
@n-riesco Updates made..

n-riesco commented:
@tarzzz LGTM. We can merge once the sample DBs are restored and all the tests pass.

I'd like reduce the noise in the history repo and set `Squash and merge` as the default option for merging (I'd like to `squash and merge` all PRs that have only commits from one author).

I don't have the rights to do this. Could you do this for me, please?

tarzzz commented:
>  We can merge once the sample DBs are restored and all the tests pass.

As mentioned in https://github.com/plotly/streambed/issues/10022#issuecomment-355343268 We need some work(1~2 days) to get test databases up and running. 
cc: @scjody 

> I don't have the rights to do this. Could you do this for me, please?

Will do.

n-riesco commented:
Once #335 is merged. I'd remove the last 3 commits (https://github.com/plotly/falcon-sql-client/pull/329/commits/a3bf08f5418091cbba92818601eaf45971dec612 , https://github.com/plotly/falcon-sql-client/pull/329/commits/76bbcea3fc438c9e5fbaf2dc3997a0703d24b790 and https://github.com/plotly/falcon-sql-client/pull/329/commits/95edf5717d5b963dbd07b95a6454ef6cbfaa1945) and rebase against master. This should fix the test failures.
